Breaking: Default to --reset behavior (fixes #2100)

btmills opened the pull request:
This is a pass at the code and (some of the) documentation updates that will fix #2100. Right now some cli-engine, cli, and config tests that use the test fixtures are failing, but it'll be cleaner to fix those once a fix for #2736 lands and we can isolate the test fixtures from the project-level .eslintrc. #2713 will be nice as well, though if it's not ready by the time this is, `"extends": "./conf/eslint.json"` will work in the interim.

lo1tuma commented on docs/user-guide/command-line-interface.md:
```diff
@@ -172,14 +171,6 @@ Example:
-
 ### `--rule`
 
 This option specifies rules to be used. These rules will be merged with the default rules and any rules specified with configuration files. (You can use `--reset` and `--no-eslintrc`, respectively, to change those behaviors.) To define multiple rules, separate them using commas, or use the flag multiple times. The [levn](https://github.com/gkz/levn#levn--) format is used for specifying the rules.
```
Can you also update this note about `--reset`.

ilyavolodin commented:
@btmills I think there's a misunderstanding about reset by default. As far as I understand, when we enable reset by default, it will reset all of the environments as well. And basically, it will not allow us to have environments turn rules on/off. From your commit it looks like there are recommended rules for node env. there. Which as far as I understand goes against what was discussed.

btmills commented:
Yeah, I wasn't totally sure how to handle that. Is the "(recommended)" annotation really just a suggestion, or is it specifically saying "this rule will be enabled by extending 'eslint:recommended'"?

btmills commented:
If the latter, then I can remove all of the environment-specific rule settings and those annotations from the index.)

ilyavolodin commented:
I believe it's the latter. @nzakas correct me if I'm wrong, please.

nzakas commented:
Yup, it's the latter.

gcochard commented:
@ilyavolodin Take another look, I'll try to get all the tests passing tonight, but the implementation is good.

btmills commented:
I just pushed an update addressing comments to this point.

I know @nzakas said #1898 would be last, but can we get an exception to remove cli before doing all of the config changes? That way we don't spend time fixing all those tests just to remove them.

btmills commented:
Or is that just that we stop exporting cli, and `bin/eslint.js` will continue to consume it internally?

ilyavolodin commented:
My understanding was that we will stop exporting cli. But maybe I'm wrong.

nzakas commented:
We just stop exporting. The file will still exist and still be used internally.

nzakas commented on lib/config.js:
```diff
@@ -302,15 +296,13 @@ function Config(options) {
-        // If `options.reset` is truthy or `options.baseConfig` is set to `false`,
-        // disable all default rules and environments
-        this.baseConfig = { rules: {} };
+    if (options.baseConfig === true) {
```
This is wrong. baseConfig will be an object if not false.

btmills replied:
Thanks, that makes so much more sense!

nzakas commented on lib/config-initializer.js:
```diff
@@ -40,6 +40,11 @@ function writeFile(config, isJson, callback) {
 function processAnswers(answers) {
     var config = {rules: {}, env: {}};
+
+    if (answers.recommended) {
```
This is already being added in #2918

btmills commented:
Rebased, tests fixed, docs updated, "WIP" removed - this is ready for thorough review before merging!

nzakas commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -397,10 +397,10 @@ The rule naming conventions for ESLint are fairly simple:
 
-* If the same rule exists in JSHint and is turned on by default, it must have the same message and be enabled by default.
-* If the same rule exists in JSLint but not in JSHint, it must have the same message and be disabled by default.
+* If the same rule exists in JSHint and is turned on by default, it must have the same message and be recommended.
```
I think we should remove this section now. We can't make any guarantees about JSHint/JSLint compat.

nzakas commented on docs/rules/README.md:
```diff
@@ -1,221 +1,221 @@
 # Rules
 
-Rules in ESLint are divided into several categories to help you better understand their value. Additionally, not all rules are enabled by default. Those that are not enabled by default are marked as being off.
+Rules in ESLint are divided into several categories to help you better understand their value. Though none are enabled by default, some are specifically recommended by ESLint.
```
..."and may be turned on using configuration."

btmills replied:
Should I specifically mention eslint:recommended somehow?

nzakas commented:
Looks good! Just a couple of documentation things.

btmills commented:
Updated the docs.

nzakas commented:
Okay, here we go!
